example.c

```diff
@@ -3,72 +3,139 @@
+}
+
+
+
 int main(int argc, char **argv) {
-    char *line;
     char *prgname = argv[0];
+    char line[1024];
 
-    /* Parse options, with --multiline we enable multi line editing. */
+    /* Parse options */
     while(argc > 1) {
         argc--;
         argv++;
-        if (!strcmp(*argv,"--multiline")) {
-            linenoiseSetMultiLine(1);
-            printf("Multi-line mode enabled.\n");
-        } else if (!strcmp(*argv,"--keycodes")) {
+        if (!strcmp(*argv,"--keycodes")) {
             linenoisePrintKeyCodes();
             exit(0);
         } else {
-            fprintf(stderr, "Usage: %s [--multiline] [--keycodes]\n", prgname);
+            fprintf(stderr, "Usage: %s [--keycodes]\n", prgname);
             exit(1);
         }
     }
 
-    /* Set the completion callback. This will be called every time the
-     * user uses the <tab> key. */
-    linenoiseSetCompletionCallback(completion);
-    linenoiseSetHintsCallback(hints);
+    printf("Press Ctrl-D or type '/quit' to quit\r\n");
+    printf("Unix users: Make sure terminal is in raw mode: eg 'stty raw -echo'\r\n");
+    printf("When you have quit, blind-type 'reset' to reset your terminal\r\n");
 
     /* Load history from file. The history file is just a plain text file
      * where entries are separated by newlines. */
-    linenoiseHistoryLoad("history.txt"); /* Load the history at startup */
+    linenoiseHistoryLoad("history.txt");
+    /* If we don't have history.txt, need at least one existing history
+    entry to set up the buffer correctly */
+    linenoiseHistoryAdd("previously-entered");
 
     /* Now this is the main loop of the typical linenoise-based application.
-     * The call to linenoise() will block as long as the user types something
-     * and presses enter.
-     *
-     * The typed string is returned as a malloc() allocated string by
-     * linenoise, so the user needs to free() it. */
-    while((line = linenoise("hello> ")) != NULL) {
-        /* Do something with the string. */
-        if (line[0] != '\0' && line[0] != '/') {
-            printf("echo: '%s'\n", line);
-            linenoiseHistoryAdd(line); /* Add to the history. */
-            linenoiseHistorySave("history.txt"); /* Save the history on disk. */
-        } else if (!strncmp(line,"/historylen",11)) {
-            /* The "/historylen" command will change the history len. */
-            int len = atoi(line+11);
-            linenoiseHistorySetMaxLen(len);
-        } else if (line[0] == '/') {
-            printf("Unreconized command: %s\n", line);
+     * linenoiseEdit won't block to allow other work to continue if no keys are pressed
+     */
+
+     int ret = 0;
+     int something_else = 0;
+     do {
+        ret = linenoiseEdit(line, sizeof(line),"hello> ");
+        /* ret is the number of characters returned,
+         * -1 that the line was incomplete, -2 for Ctrl-D was pressed.
+         * Do something with the returned string if valid. */
+        if (ret > 0) {
```

Comment:
Avoid deep nesting of conditionals.

Reply:
Thanks for the comments. Can you explain how you would avoid that in a clean way here? We can't break out of the loop because there is work to do at the end (and I want to maintain the style to illustrate that this is a small part of a bigger work loop).  If we pushed the `if (ret>0)` test into every conditional, that would become more unwieldy, would it not? We need the conditionals to be exclusive, but they aren't amenable to a switch statement. So what alternative would you suggest?

Reply:
The way you are handling this right now `something_else++` is executed in any case, even if there is some action taken by the loop code. One way you can avoid the extra level is to leverage the `if`/`else if` we already have, by simply making the inverse check the first block and then do nothing:

```c
if (ret <= 0) {
  // No new input data
} else if (line[0] != 0 && line[0] != '/') {
  // And so forth …
}

something_else++;
```

Reply:
Ah, OK. Not sure it makes it hugely clearer but fair enough.  I've refactored that and commented the empty `(ret <= 0)` clause so it explains what cases that covers